pyriemann/embedding.py

```diff
@@ -96,9 +102,218 @@ def fit_transform(self, X, y=None):
+    v = np.ones(n_neighbors, dtype=X.dtype)
+
+    for i in range(len(X)):
+        X_neighbors = Y[indices[i]]
+        G = kernel_riemann(X_neighbors, Cref=X[i])
+        trace = np.trace(G)
+        if trace > 0:
+            R = reg * trace
+        else:
+            R = reg
+        G.flat[:: n_neighbors + 1] += R
+        w = solve(G, v, sym_pos=True)
+        B[i, :] = w / np.sum(w)
+    return B
+
+
+def riemann_lle(X,
+                n_components=2,
+                n_neighbors=5,
+                reg=1e-3,
+                null_space_args={}):
+    """Riemannian Locally Linear Embedding (LLE).
+
+    Riemannian LLE as proposed in [1]_. LLE is a non-linear, neighborhood-
+    preserving dimensionality reduction algorithm which consists of three
+    main steps:
+    For each point x,
+    1.  find its k nearest neighbors KNN(x) and
+    2.  calculate the best reconstruction of x based on its KNN.
+    3.  Then calculate a low-dimensional embedding for all points based on
+        the weights in step 2.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_matrices, n_channels, n_channels)
+        ndarray of SPD matrices.
+    n_components : int, default: 2
+        Dimensionality of projected space.
+    n_neighbors : int, default: 5
+        Number of neighbors for reconstruction of each point.
+    n_jobs : int, default: 1
+        The number of jobs to use for the computation. This works by computing
+        each of the class centroid in parallel.
+        If -1 all CPUs are used. If 1 is given, no parallel computing code is
+        used at all, which is useful for debugging. For n_jobs below -1,
+        (n_cpus + 1 + n_jobs) are used. Thus for n_jobs = -2, all CPUs but one
+        are used.
+    reg : float, default: 1e-3
+        Regularization parameter.
+    null_space_args : dict, default {}
+        Keyword arguments passed to
+        sklearn.manifold._locally_linear.null_space.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+    See sklearn.manifold._locally_linear.locally_linear_embedding for original
+    code.
+
+    References
+    ----------
+    .. [1] A. Goh and R. Vidal, "Clustering and dimensionality reduction
+        on Riemannian manifolds", 2008 IEEE Conference on Computer Vision
+        and Pattern Recognition, June 2008
+    """
+
+    n_matrices, n_channels, n_channels = X.shape
+    pairwise_distances = pairwise_distance(X, metric='riemann')
+    neighbors = np.array([np.argsort(dist)[1:n_neighbors + 1]
+                          for dist in pairwise_distances])
+
+    B = barycenter_weights(X, X, neighbors, reg=reg)
+
+    indptr = np.arange(0, n_matrices * n_neighbors + 1, n_neighbors)
+    W = csr_matrix((B.ravel(), neighbors.ravel(), indptr), shape=(n_matrices,
+                                                                  n_matrices))
+    M = (W.T * W - W.T - W).toarray()
+    M.flat[:: M.shape[0] + 1] += 1  # W = W - I = W - I
```

Comment:
here you add 1 to M and not remove 1 to W

Reply:
that's how they did it in sklearn and the two lines should do just what is asked.

Reply:
it does not mean it should copy pasted :)

disclaimer I wrote a quiet some part of sklearn myself so I can criticise :)

Reply:
Seems that the comment is wrong. it should be
M = (W - I).T@(W - I) 
which boils down to 
W.T@W - W - W.T + I
which is exactly what they did although I don't know why they write it up so weirdly. I'll change it just so it's more readable.

Reply:
feel free to send a PR on scikit-learn to fix the comment there too 🙏

Reply:
will do!